### PR TITLE
Support custom directives in InMemory Cache

### LINKED ISSUE #2709: Support custom directives in InMemory Cache
**Intended outcome:**
I would like to add custom directive support to "apollo-cache-inmemory". More precisely fix it to prevent corrupted cache.

If you are not familiar with custom directives you can checkout my new module: [graphql-directive](https://github.com/smooth-code/graphql-directive).

**Actual outcome:**
"apollo-cache-inmemory" relies on [`getStoreKeyName` function](https://github.com/apollographql/apollo-client/blob/master/packages/apollo-utilities/src/storeUtils.ts#L167) either for writing or for reading cache. By reading the code of this method you can see that a field `firstName @upperCase` and `firstName` will use the same cache key: "firstName". Even if the value is not the same, it corrupts cache.

**How to reproduce the issue:**
```js
const gql = require('graphql-tag')
const { InMemoryCache } = require('apollo-cache-inmemory')

const cache = new InMemoryCache()

const query = gql`
  query Foo {
    book(id: 10) {
      __typename
      name @uppercase
    }
  }
`

cache.writeQuery({
  query,
  data: {
    book: {
      __typename: 'BOOK',
      name: 'harry',
    },
  },
})

console.log(cache.extract())
console.log(cache.readQuery({ query }))
```

**Version**
- apollo-cache-inmemory@1.1.4
- apollo-utilities@1.0.3

### PR DESCRIPTION
**Pull request by neoziro**
Before this change, using a custom directive corrupted the cache.

It is now possible to safely use a custom directive without breaking it.

Example of custom directives: https://github.com/smooth-code/graphql-directive

Closes #2709

### CODE REVIEW

File: packages/apollo-utilities/src/storeUtils.ts [hunk 1]
```diff
@@ -197,13 +199,25 @@ export function getStoreKeyName(
...
     const stringifiedArgs: string = JSON.stringify(args);
+    completeFieldName += `(${stringifiedArgs})`;
+  }
 
-    return `${fieldName}(${stringifiedArgs})`;
+  if (directives) {
+    Object.keys(directives).forEach(key => {
+      if (KNOWN_DIRECTIVES.indexOf(key) !== -1) return;
+      if (directives[key] && Object.keys(directives[key]).length) {
+        completeFieldName += `@${key}(${JSON.stringify(directives[key])})`;
```
Comment (giautm) [inline packages/apollo-utilities/src/storeUtils.ts]: Hey, how about directives arguments? Should put it in the key too?

Comment (neoziro) [reply packages/apollo-utilities/src/storeUtils.ts]: Actually they are! `directives[key]` contains arguments.

**Comment by jbaxleyiii** [comment]
@neoziro NICE! This is great! Can you fix the conflicts in the authors file and rebase on master? Happy to get this in!

**Comment by neoziro** [comment]
@jbaxleyiii done

**Comment by neoziro** [comment]
@jbaxleyiii the CI seems broken, `lerna bootstrap` is failing.

**Comment by jbaxleyiii** [comment]
@neoziro odd! Can you rebase this on top of master? That should fix it and I can merge in for the next release!

**Comment by neoziro** [comment]
@jbaxleyiii done

**Comment by neoziro** [comment]
@jbaxleyiii can you merge it?

File: packages/apollo-utilities/src/storeUtils.ts (hunk 1 above)
Comment (fbartho) [reply packages/apollo-utilities/src/storeUtils.ts]: Should the keys in `JSON.stringify(directives[key])` be pre-sorted into a consistent order? Right now, they'll hash in undefined order, no?

**Comment by neoziro** [comment]
@peggyrayzis do we have a solution to make it works? Also it looks like a test was missing in `apollo-link-state`.

**Pull request merged**